### PR TITLE
Integrate notification banner component into Compare View (NDDB)

### LINKED ISSUE #4767: Display a notification on History tab when the base branch moves ahead of the current branch (NDDB)
We need a visual cue to notify users that their current branch has fallen behind its base branch. The visual cue should be a small blue dot placed next to the label on the History tab. 

### Notification Rules

Tl;dr: show a dot whenever the base branch has commits the current branch doesn't and hide it whenever the History tab is clicked.

**Showing notification**
* The current branch falls behind the base branch (number commits behind is more than zero)
* The current branch was already behind the base branch because the user consumed the last notification without performing any actions and has fallen even more behind (number of commits behind is more than 0 **and** different from the last time the notification was shown)

**Removing notification**
* User clicks on the notification (History tab)

### PR DESCRIPTION
**Pull request by iAmWillShepherd**
🌵 🌵 🌵 This PR depends on PRs ~#4794~ and ~#4783~ being merged first🌵 🌵 🌵 

Fixes #4767

This PR is just a continuation of the work done in ☝️ and integrates it into the _Compare view_ UI.

See the gifs 👇 to see it in action.

<details>
<summary>Testing</summary>
If you notice the banner doesn't popup right away, try clicking on another tab or performing a manual fetch ([more info](https://github.com/desktop/desktop/projects/10#card-10732873))

### Unpublished repository
1. Create a new branch (it does not matter if you branch off `master` or another branch)
2. Make some commits on `master`
3. Checkout the new branch you created
4. Notification dot and banner should be visible
5. Comparison branch should be `master`

### Published repository (non-fork)
1. Create and publish a new branch (it does not matter if you branch off the _default branch_ or another branch)
2. Make some commits on the _default branch_ and push
3. Checkout the new branch you created
4. Notification dot and banner should be visible
5. Comparison branch should be _default branch_ (likely master)

### Published repository (non-fork) with PR
1. Create a new pull request targeting a branch that’s not the _default branch_
2. Make some commits on the _target branch_
3. Checkout the PR branch 
4. Notification dot and banner should be visible
5. Comparison branch should be the _target branch_

### Published repository (fork) - Starting point for new project (origin/default-branch is ahead of current branch)
1. Fork a repo and clone the forked repo
2. Create a new branch
3. Make some commits on the _default branch_
3. Checkout the branch you created
4. Notification dot and banner should be visible
5. Comparison branch should be the _default branch_ of the forked repo

### Published repository (fork) - Contributing to another project (origin/default-branch is not ahead of current branch)
1. Fork a repo and clone the forked repo
2. Create a new branch
3. Make some commits on the parent repo’s _default branch_
3. Perform a manual fetch (or wait, your choice)
4. Notification dot and banner should be visible
5. Comparison branch should be the _default branch_ of the parent repo (likely `upstream/master`)

</details>

### CODE REVIEW

**Comment by lakegirl5588** [comment]
PLEASE STOP SENDING EMAILS

THANKS,

MARIA BALLARD

File: app/src/ui/app.tsx [hunk 1]
```diff
@@ -1651,26 +1651,27 @@ export class App extends React.Component<IAppProps, IAppState> {
     }
 
     if (selectedState.type === SelectionType.Repository) {
-      const externalEditorLabel = this.state.selectedExternalEditor
+      const externalEditorLabel = state.selectedExternalEditor
```
Comment (iAmWillShepherd) [inline app/src/ui/app.tsx]: These changes are really just a bit of cleanup. We had a variable holding state, so I just removed `this` from each usage of `state`.

File: app/src/ui/app.tsx [hunk 2]
```diff
@@ -1651,26 +1651,27 @@ export class App extends React.Component<IAppProps, IAppState> {
...
+          imageDiffType={state.imageDiffType}
           askForConfirmationOnDiscardChanges={
-            this.state.askForConfirmationOnDiscardChanges
+            state.askForConfirmationOnDiscardChanges
           }
-          accounts={this.state.accounts}
+          accounts={state.accounts}
           externalEditorLabel={externalEditorLabel}
           onOpenInExternalEditor={this.openFileInExternalEditor}
+          isDivergingBranchBannerVisible={state.isDivergingBranchBannerVisible}
```
Comment (iAmWillShepherd) [inline app/src/ui/app.tsx]: This is the only line I actually added.

File: app/src/ui/notification/new-commits-banner.tsx [hunk 3]
```diff
@@ -14,6 +14,8 @@ interface INewCommitsBannerProps {
    * from the current branch
    */
   readonly baseBranch: Branch
+
```
Comment (shiftkey) [inline app/src/ui/notification/new-commits-banner.tsx]: 📝

File: app/src/ui/repository.tsx [hunk 4]
```diff
@@ -51,8 +51,9 @@ interface IRepositoryViewProps {
    *
    * @param fullPath The full path to the file on disk
    */
-
   readonly onOpenInExternalEditor: (fullPath: string) => void
+
```
Comment (shiftkey) [inline app/src/ui/repository.tsx]: :memo:

File: app/src/lib/stores/app-store.ts [hunk 5]
```diff
@@ -268,6 +270,8 @@ export class AppStore extends TypedBaseStore<IAppState> {
   private selectedBranchesTab = BranchesTab.Branches
   private selectedTheme = ApplicationTheme.Light
 
+  private isDivergingBranchBannerVisible: boolean = false
```
Comment (shiftkey) [inline app/src/lib/stores/app-store.ts]: :art: no need to `: boolean` here

File: app/src/lib/stores/helpers/infer-comparison-branch.ts [hunk 6]
```diff
@@ -98,17 +94,27 @@ async function getDefaultBranchOfFork(
...
+    ghRepo.parent !== null &&
+    ghRepo.parent.defaultBranch !== null
+  ) {
     return defaultBranch
   }
 
-  if (ghRepo.parent !== null && ghRepo.parent.defaultBranch !== null) {
-    return getDefaultBranchOfForkedGitHubRepo(repository, branches, getRemotes)
-  }
+  const potentialdefaultBranchOfParent = await getDefaultBranchOfForkedGitHubRepo(
```
Comment (shiftkey) [inline app/src/lib/stores/helpers/infer-comparison-branch.ts]: :art: `potentialDefault...`

File: app/src/lib/stores/app-store.ts [hunk 7]
```diff
@@ -2942,6 +2994,11 @@ export class AppStore extends TypedBaseStore<IAppState> {
     this.emitUpdate()
   }
 
+  public _setDivergingBranchBannerVisibility(visibility: boolean) {
```
Comment (shiftkey) [inline app/src/lib/stores/app-store.ts]: :art: `visible` to be consistent with `isDivergingBranchBannerVisible`

File: app/src/lib/stores/helpers/infer-comparison-branch.ts [hunk 8]
```diff
@@ -98,17 +94,27 @@ async function getDefaultBranchOfFork(
     return getMasterBranch(branches)
   }
 
-  const range = revRange(currentBranch.tip.sha, defaultBranch.tip.sha)
-  const aheadBehind = await getAheadBehind(repository, range)
-  if (aheadBehind !== null && aheadBehind.ahead > 0) {
+  // I need to keep track of ahead/behind of two branches and return based on which one has ahead > 0
```
Comment (shiftkey) [inline app/src/lib/stores/helpers/infer-comparison-branch.ts]: :art: can this comment be descriptive about what it's actually looking for below?

File: app/src/ui/notification/new-commits-banner.tsx [hunk 9]
```diff
@@ -14,6 +14,7 @@ interface INewCommitsBannerProps {
    * from the current branch
    */
   readonly baseBranch: Branch
+  readonly onDismiss: () => void
```
Comment (shiftkey) [inline app/src/ui/notification/new-commits-banner.tsx]: 📝

File: app/test/unit/infer-comparison-branch-test.ts [hunk 10]
```diff
@@ -84,8 +78,14 @@ describe('inferComparisonBranch', () => {
     createTestBranch('default', '3', 'origin'),
     createTestBranch('head', '4', 'origin'),
     createTestBranch('upstream/base', '5', 'upstream'),
+    createTestBranch('fork', '6', 'origin'),
   ]
 
+  let mockComparisonCache: ComparisonCache
+  beforeEach(() => {
+    mockComparisonCache = new ComparisonCache()
```
Comment (shiftkey) [inline app/test/unit/infer-comparison-branch-test.ts]: :art: a _mock_ object is one that simulates the behaviour of a real method or object. This looks like you're using the same implementation, so could this just be named `comparisonCache`?

**Comment by iAmWillShepherd** [comment]
🥇

**Comment by shiftkey** [comment]
Nice work @iAmWillShepherd!

**Pull request merged**